core/src/wallet/ethereum/ERC20/ERC20LikeAccount.cpp

```diff
@@ -72,15 +73,72 @@ namespace ledger {
         std::shared_ptr<api::BigInt> ERC20LikeAccount::getBalance() {
             auto result = BigInt::ZERO;
             auto operations = getOperations();
+
             for (auto &op : operations) {
-                if (op->getOperationType() == api::OperationType::RECEIVE) {
-                    result = result + BigInt(op->getValue()->toString(10));
-                } else if (op->getOperationType() == api::OperationType::SEND){
-                    result = result - BigInt(op->getValue()->toString(10));
+                result = accumulateBalanceWithOperation(result, *op);
+            }
+
+            return std::make_shared<api::BigIntImpl>(result);
+        }
+
+        api::ERC20LikeBalanceHistory ERC20LikeAccount::getBalanceHistoryFor(
+            const std::chrono::system_clock::time_point& startDate,
+            const std::chrono::system_clock::time_point& endDate,
+            api::TimePeriod period
+        ) {
+            // guard against bad arguments
+            if (startDate >= endDate) {
+                throw make_exception(
+                    api::ErrorCode::INVALID_DATE_FORMAT,
+                    "Start date should be strictly greater than end date"
+                );
+            }
+
+            std::vector<std::shared_ptr<api::BigInt>> balances;
+            auto currentBalance = BigInt::ZERO;
+            auto nextDate = DateUtils::incrementDate(startDate, period);
+            auto operations = getSortedOperationsRange(startDate, endDate);
```

Comment:
You will always have Balance 0 at startDate

Reply:
Yes. (I don’t know what else to do, is there a problem with that? :D)

Reply:
I think this is not users expect, we should calculate what is the balance at startPoint as the first element of balances array. Either it is one request to DB or manually - in the same loop.

Reply:
If you compute the balance at `startPoint`, you’re actually not using `startPoint` because it means `startPoint = 0`. I don’t really see why Live would need to compute the balance with a `startPoint` actually (they only know the `endDate` to me, which represents the pointer at which the balance should be displayed). Actually, with hindsight, I don’t really get neither of the solution. I’m confused. :D

Reply:
Talked on slack, I’ll fix that. :)

Reply:
@teams2ua that should be fixed now. Thanks for spotting that issue out!